File: wrappers/wasm/tests/foundation/Sha256.test.js
const initFoundation = require('../../dist/foundation/node.cjs');
const { hexToUint8Array } = require('../utils');

describe('Sha256', () => {
  let foundation;
  let sha256;

  beforeEach(async () => {
    foundation = await initFoundation();
    sha256 = new foundation.Sha256();
  });

  describe('hash', () => {
    test('1', () => {
      const input = new Uint8Array();
      const result = sha256.hash(input);
      const expectedResult = hexToUint8Array('e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855');
      expect(result.toString()).toBe(expectedResult.toString());
    });

    test('2', () => {
      const input = hexToUint8Array('bd');
      const result = sha256.hash(input);
      const expectedResult = hexToUint8Array('68325720aabd7c82f30f554b313d0570c95accbb7dc4b5aae11204c08ffe732b');
      expect(result.toString()).toBe(expectedResult.toString());
    });

    test('3', () => {
      const input = hexToUint8Array('5fd4');
      const result = sha256.hash(input);
      const expectedResult = hexToUint8Array('7c4fbf484498d21b487b9d61de8914b2eadaf2698712936d47c3ada2558f6788');
      expect(result.toString()).toBe(expectedResult.toString());
    });
  });

  describe('stream', () => {
    test('1', () => {
      const input = new Uint8Array();
      sha256.start();
      sha256.update(input);
      const result = sha256.finish();
      const expectedResult = hexToUint8Array('e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855');
      expect(result.toString()).toBe(expectedResult.toString());
    });

    test('2', () => {
      const input = hexToUint8Array('bd');
      sha256.start();
      sha256.update(input);
      const result = sha256.finish();
      const expectedResult = hexToUint8Array('68325720aabd7c82f30f554b313d0570c95accbb7dc4b5aae11204c08ffe732b');
      expect(result.toString()).toBe(expectedResult.toString());
    });

    test('3', () => {
      const input = hexToUint8Array('5fd4');
      sha256.start();
      sha256.update(input);
      const result = sha256.finish();
      const expectedResult = hexToUint8Array('7c4fbf484498d21b487b9d61de8914b2eadaf2698712936d47c3ada2558f6788');
      expect(result.toString()).toBe(expectedResult.toString());
    });
  });
});

File: wrappers/python/manual/virgil_crypto/tests/foundation/sha256_test.py
import unittest
from base64 import b64encode
from binascii import unhexlify

from virgil_crypto.foundation.sha256 import SHA256


class Sha256Test(unittest.TestCase):

    VECTOR_1_INPUT_BYTES = bytearray()
    VECTOR_1_DIGEST_BYTES = unhexlify("E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855")

    VECTOR_2_INPUT_BYTES = unhexlify("BD")
    VECTOR_2_DIGEST_BYTES = unhexlify("68325720AABD7C82F30F554B313D0570C95ACCBB7DC4B5AAE11204C08FFE732B")

    VECTOR_3_INPUT_BYTES = unhexlify("5FD4")
    VECTOR_3_DIGEST_BYTES = unhexlify("7C4FBF484498D21B487B9D61DE8914B2EADAF2698712936D47C3ADA2558F6788")

    def test_sha256_hash_empty_string_success(self):
        sha_hash = SHA256().hash(bytearray("".encode()))
        self.assertEqual("47DEQpj8HBSa+/TImW+5JCeuQeRkm5NMpJWZG3hSuFU=", b64encode(sha_hash).decode())

    def test_hash_empty_bytes(self):
        sha256 = SHA256()
        empty_bytes_hash = sha256.hash(self.VECTOR_1_INPUT_BYTES)
        self.assertEqual(
            empty_bytes_hash,
            self.VECTOR_1_DIGEST_BYTES
        )

    def test_vector_2(self):
        sha256 = SHA256()
        data_hash = sha256.hash(self.VECTOR_2_INPUT_BYTES)
        self.assertEqual(
            data_hash,
            self.VECTOR_2_DIGEST_BYTES
        )

    def test_vector_3(self):
        sha256 = SHA256()
        data_hash = sha256.hash(self.VECTOR_3_INPUT_BYTES)
        self.assertEqual(
            data_hash,
            self.VECTOR_3_DIGEST_BYTES
        )

    # Test implementation for hash stream

    def test_hash_stream_vector_1(self):
        sha256 = SHA256()

        sha256.start()
        sha256.update(self.VECTOR_1_INPUT_BYTES)
        digest = sha256.finish()

        self.assertEqual(len(self.VECTOR_1_DIGEST_BYTES), len(digest))
        self.assertEqual(self.VECTOR_1_DIGEST_BYTES, digest)

    def test_hash_stream_vector_2(self):
        sha256 = SHA256()
        sha256.start()
        sha256.update(self.VECTOR_2_INPUT_BYTES)
        digest = sha256.finish()

        self.assertEqual(len(self.VECTOR_2_DIGEST_BYTES), len(digest))
        self.assertEqual(self.VECTOR_2_DIGEST_BYTES, digest)

    def test_hash_stream_vector_3(self):
        sha256 = SHA256()

        sha256.start()
        sha256.update(self.VECTOR_3_INPUT_BYTES)
        digest = sha256.finish()

        self.assertEqual(len(self.VECTOR_3_DIGEST_BYTES), len(digest))
        self.assertEqual(self.VECTOR_3_DIGEST_BYTES, digest)